sharktank/sharktank/kernels/gemm_fp4_asm.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2025 Advanced Micro Devices, Inc.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from sharktank.kernels.base import *
+from sharktank.kernels.mlir_kernel import *
+import torch
+
+
+M = DynDim.M
+N = DynDim.N
+K = DynDim.K
+HALF_K = DynDim.HALF_K
+K_OVER_THIRTYTWO = DynDim.K_OVER_THIRTYTWO
+
+U8 = Dtype.U8(torch.uint8)
+# F8 = Dtype.F8(torch.float8_e8m0fnu)
+F32 = Dtype.F32(torch.float32)
+BF16 = Dtype.BF16(torch.bfloat16)
+
+
+@mlir_kernel(
+    inputs=(
+        MLIRTensor[M, HALF_K, U8],
+        MLIRTensor[N, HALF_K, U8],
+        MLIRTensor[M, K_OVER_THIRTYTWO, U8],
+        MLIRTensor[N, K_OVER_THIRTYTWO, U8],
+        MLIRTensor[M, N, F32],
+    ),
+    results=(MLIRTensor[M, N, F32],),
+)
+def asm_fp4_gemm(x, w, x_scale, w_scale, bias, result=None):
```

Comment:
Could you add comments to this MLIR to group logical subsections for better readability?

Reply:
It could also benefit from variables being renamed to meaningful things, such as `%arg7`